### PR TITLE
126-fetch-lecture-streams

### LINKED ISSUE #126: Fetch lecture streams
Replace stateful approach with stateless approach where we call the stream endpoints and pass the result through the views, e.g.:
- User clicks on course card -> CourseDetail view is created and gets passed result of GetCoursesStreams
- User clicks on a stream -> VideoCard is created and gets passed stream as argument

### LINKED ISSUE #170: Add automatic progress saving to videoPlayer

### LINKED ISSUE #171: Add switching between different playlist sources
## Description 

- Add switching betwwen `playlistUrl`, `playlistUrlCAM` and `playlistUrlPRES`.

### PR DESCRIPTION
**Pull request by GravityDarkLab**
## Demo

https://github.com/TUM-Dev/gocast_mobile/assets/101757413/df8c8848-800a-4d5f-8a8b-1509ade1a166

## Add

- A custom menu Bar under the videoPlayer

## Features

- Video fetching
- Dynamic SearchBar icon
- Video progress saving
- switching between combined, camera and presentation views

## Fixes

- Replaced `pinned_courses_base_view` with the existing `base_view`

## Resolves

- closes #126 
- closes #170 
- closes #171

### CODE REVIEW

**Comment by Anishyou** [comment]
Looks good to me. But would like to test with actual data.

**Comment by GravityDarkLab** [comment]
> Looks good to me. But would like to test with actual data.
> 
> 

It is still A Draft as mentioned and not complete yet

**Comment by ahmetsenturk** [comment]
I logged in with the `studi1` credentials and could only see dummy videos. Are we fetching some actual lectures from live.rgb as well? If so how can I test it?

**Comment by GravityDarkLab** [comment]
> I logged in with the `studi1` credentials and could only see dummy videos. Are we fetching some actual lectures from live.rgb as well? If so how can I test it?

Ensure you have docker running and api in this branche with latest changes: https://github.com/TUM-Dev/gocast/tree/IPraktikum-dev

**Review by milenark — COMMENTED** [review]
Looks pretty good so far - just changing the playback speeds is not working yet, are we adding that later or does it still need to be changed here?

File: lib/views/video_view/custom_video_control_bar.dart [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                onPressed: onToggleChat,
+              ),
+              IconButton(
+                icon: const Icon(Icons.quiz_outlined),
+                onPressed: onOpenQuizzes,
+              ),
+            ],
+          ),
+
+          // Three-dot menu on the right
```
Comment (carlobortolan) [inline lib/views/video_view/custom_video_control_bar.dart:40]: I'd suggest only showing the icons of the avaiable views (e.g., if a stream only has presentation view (and therefore only that URL is set and all other URL-fields are either empty or null, only the presentation view icon is shown)

File: lib/views/video_view/custom_video_control_bar.dart [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+          PopupMenuButton<String>(
+            onSelected: (choice) => onMenuSelection(choice, currentStream),
+            itemBuilder: (BuildContext context) {
+              return <Map<String, IconData>>[
```
Comment (carlobortolan) [inline lib/views/video_view/custom_video_control_bar.dart:44]: We need also the 'Split view' icon which is shown whenever both presentation and camera view are available (= the view where you can manually resize the presentation and camera view next to each other)

**Review by carlobortolan — CHANGES_REQUESTED** [review]
Works great :+1: 
I only found the following bug (on Android) when pressing "Praktikum-Golang" -> "Intro to Go" -> "Camera View"
image

**Comment by carlobortolan** [comment]
> Looks pretty good so far - just changing the playback speeds is not working yet, are we adding that later or does it still need to be changed here?

I tried it out on Android and there it works as it should; can you maybe attach a screenshot/video where this happens?

**Comment by GravityDarkLab** [comment]
> ssing "Praktikum-Golan

I added more checks now in the `_initVideoPlayer()`, also i found that the `_completionListener` wasn´t being disposed and he was probably causing the root problem

File: lib/views/video_view/custom_video_control_bar.dart (hunk 1 above)
Comment (GravityDarkLab) [reply lib/views/video_view/custom_video_control_bar.dart:57]: Done ✅ 
It was also causing the previous issue.

**Comment by GravityDarkLab** [comment]
> Looks pretty good so far - just changing the playback speeds is not working yet, are we adding that later or does it still need to be changed here?

@milenark 

I tested again on both iPhone and android but was working.
Can you provide more details?

File: lib/views/video_view/custom_video_control_bar.dart (hunk 2 above)
Comment (GravityDarkLab) [reply lib/views/video_view/custom_video_control_bar.dart]: Done ✅

**Comment by Anishyou** [comment]
Looks all good to me :). Great work man! I will have a look for android and get back to u

**Comment by milenark** [comment]
> I tested again on both iPhone and android but it was working.

I reinstalled everything and it works for me now too, my bad

**Comment by carlobortolan** [comment]
> I added more checks now in the `_initVideoPlayer()`, also i found that the `_completionListener` wasn´t being disposed and he was probably causing the root problem

Nevermind, you can ignore this :sweat_smile:; I had a closer look at it and as it turns out it was an error with my database which had the wrong PlaylistURL set (see screenshot - for testing purposes I changed all playlist_url_cam and forgot to change it back)
Here' s the error message:
```
E/ExoPlayerImplInternal( 2048): Playback error
E/ExoPlayerImplInternal( 2048):   com.google.android.exoplayer2.ExoPlaybackException: Source error
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleIoException(ExoPlayerImplInternal.java:644)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:616)
E/ExoPlayerImplInternal( 2048):       at android.os.Handler.dispatchMessage(Handler.java:102)
E/ExoPlayerImplInternal( 2048):       at android.os.Looper.loopOnce(Looper.java:205)
E/ExoPlayerImplInternal( 2048):       at android.os.Looper.loop(Looper.java:294)
E/ExoPlayerImplInternal( 2048):       at android.os.HandlerThread.run(HandlerThread.java:67)
E/ExoPlayerImplInternal( 2048):   Caused by: com.google.android.exoplayer2.upstream.HttpDataSource$CleartextNotPermittedException: Cleartext HTTP traffic not permitted. See https://developer.android.com/guide/topics/media/issues/cleartext-not-permitted
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:382)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DefaultDataSource.open(DefaultDataSource.java:263)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.StatsDataSource.open(StatsDataSource.java:84)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DataSourceInputStream.checkOpened(DataSourceInputStream.java:99)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DataSourceInputStream.open(DataSourceInputStream.java:62)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.ParsingLoadable.load(ParsingLoadable.java:174)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.Loader$LoadTask.run(Loader.java:412)
E/ExoPlayerImplInternal( 2048):       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
E/ExoPlayerImplInternal( 2048):       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
E/ExoPlayerImplInternal( 2048):       at java.lang.Thread.run(Thread.java:1012)
E/ExoPlayerImplInternal( 2048):   Caused by: java.io.IOException: Cleartext HTTP traffic to localhost not permitted
E/ExoPlayerImplInternal( 2048):       at com.android.okhttp.HttpHandler$CleartextURLFilter.checkURLPermitted(HttpHandler.java:127)
E/ExoPlayerImplInternal( 2048):       at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:462)
E/ExoPlayerImplInternal( 2048):       at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:131)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DefaultHttpDataSource.makeConnection(DefaultHttpDataSource.java:653)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DefaultHttpDataSource.makeConnection(DefaultHttpDataSource.java:550)
E/ExoPlayerImplInternal( 2048):       at com.google.android.exoplayer2.upstream.DefaultHttpDataSource.open(DefaultHttpDataSource.java:376)
E/ExoPlayerImplInternal( 2048):       ... 9 more
I/flutter ( 2048): AppError: Failed to load video
I/flutter ( 2048): Original error: PlatformException(VideoError, Video player had error com.google.android.exoplayer2.ExoPlaybackException: Source error, null, null)
```

image

**Review by carlobortolan — APPROVED** [review]

**Review by Anishyou — APPROVED** [review]
works everything

**Pull request merged**